Split the .kibana saved objects index into multiple indices

gsoldevila opened the pull request:
## Description 

Fix https://github.com/elastic/kibana/issues/104081

This PR move some of the SO types from the `.kibana` index into the following ones:
- `.kibana_alerting_cases`
- `.kibana_analytics`
- `.kibana_security_solution`
- `.kibana_ingest`

This split/reallocation will occur during the `8.8.0` Kibana upgrade (*meaning: from any version older than `8.8.0` to any version greater or equal to `8.8.0`*)

**This PR main changes are:**
- implement the changes required in the SO migration algorithm to support this reallocation 
- update the FTR tools (looking at you esArchiver) to support these new indices
- update hardcoded references to `.kibana` and usage of the `core.savedObjects.getKibanaIndex()` to use new APIs to target the correct index/indices
- update FTR datasets, tests and utility accordingly 

## To reviewers

**Overall estimated risk of regressions: low**

But, still, please take the time to review changes in your code. The parts of the production code that were the most impacted are the telemetry collectors, as most of them were performing direct requests against the `.kibana` index, so we had to adapt them. Most other contributor-owned changes are in FTR tests and datasets.

If you think a type is misplaced (either we missed some types that should be moved to a specific index, or some types were moved and shouldn't have been) please tell us, and we'll fix the reallocation either in this PR or in a follow-up.

## .Kibana split

The following new indices are introduced by this PR, with the following SO types being moved to it. (any SO type not listed here will be staying in its current index)

Note: The complete **_type => index_** breakdown is available in [this spreadsheet](https://docs.google.com/spreadsheets/d/1b_MG_E_aBksZ4Vkd9cVayij1oBpdhvH4XC8NVlChiio/edit#gid=145920788).

#### `.kibana_alerting_cases`
- action
- action_task_params
- alert
- api_key_pending_invalidation
- cases
- cases-comments
- cases-configure
- cases-connector-mappings
- cases-telemetry
- cases-user-actions
- connector_token
- rules-settings
- maintenance-window

#### `.kibana_security_solution`
- csp-rule-template
- endpoint:user-artifact
- endpoint:user-artifact-manifest
- exception-list
- exception-list-agnostic
- osquery-manager-usage-metric
- osquery-pack
- osquery-pack-asset
- osquery-saved-query
- security-rule
- security-solution-signals-migration
- siem-detection-engine-rule-actions
- siem-ui-timeline
- siem-ui-timeline-note
- siem-ui-timeline-pinned-event

#### `.kibana_analytics`

- canvas-element
- canvas-workpad-template
- canvas-workpad
- dashboard
- graph-workspace
- index-pattern
- kql-telemetry
- lens
- lens-ui-telemetry
- map
- search
- search-session
- search-telemetry
- visualization

#### `.kibana_ingest`

- epm-packages
- epm-packages-assets
- fleet-fleet-server-host
- fleet-message-signing-keys
- fleet-preconfiguration-deletion-record
- fleet-proxy
- ingest_manager_settings
- ingest-agent-policies
- ingest-download-sources
- ingest-outputs
- ingest-package-policies

## Tasks / PRs

### Sub-PRs

**Implementation**
- 🟣 https://github.com/elastic/kibana/pull/154846
- 🟣 https://github.com/elastic/kibana/pull/154892
- 🟣 https://github.com/elastic/kibana/pull/154882
- 🟣 https://github.com/elastic/kibana/pull/154884
- 🟣 https://github.com/elastic/kibana/pull/155155

**Individual index split**
- 🟣 https://github.com/elastic/kibana/pull/154897
- 🟣 https://github.com/elastic/kibana/pull/155129
- 🟣 https://github.com/elastic/kibana/pull/155140
- 🟣 https://github.com/elastic/kibana/pull/155130

### Improvements / follow-ups 

- 👷🏼 Extract logic into [runV2Migration](https://github.com/elastic/kibana/pull/154151#discussion_r1158470566) @gsoldevila
- Make `getCurrentIndexTypesMap` resillient to intermittent failures https://github.com/elastic/kibana/pull/154151#discussion_r1169289717
- 🚧 Build a more structured [MigratorSynchronizer](https://github.com/elastic/kibana/pull/154151#discussion_r1158469918)
- 🟣 https://github.com/elastic/kibana/pull/155035
- 🟣 https://github.com/elastic/kibana/pull/155116
- 🟣 https://github.com/elastic/kibana/pull/155366
## Reallocation tweaks

Tweaks to the reallocation can be done after the initial merge, as long as it's done before the public release of 8.8

- `url` should get back to `.kibana` (see [comment](https://github.com/elastic/kibana/pull/154888#discussion_r1172317133))

## Release Note

For performance purposes, Kibana is now using more system indices to store its internal data.

The following system indices will be created when upgrading to `8.8.0`:

- `.kibana_alerting_cases`
- `.kibana_analytics`
- `.kibana_security_solution`
- `.kibana_ingest`

gsoldevila committed:
[dot-kibana-split] Allow relocating SO to different indices during migration (#154846)

Clone of https://github.com/elastic/kibana/pull/154151
Baseline for the `dot-kibana-split` feature branch

gsoldevila committed:
Merge branch 'main' into dot-kibana-split

gsoldevila committed:
[dot-kibana-split] Update dynamic mappings.json to include the `indexTypesMap` (#154892)

Part of https://github.com/elastic/kibana/pull/154888

Some mappings.json use a special `$KIBANA_PACKAGE_VERSION` variable.
ES archiver uses this variable to generate saved object indices with
names that match the current stack version.
Note that this a bit brittle, cause incompatible mappings changes would
break Kibana.
However, it allows skipping reindexing and saving some precious CI time
when running tests.

With the introduction of the .kibana split, the mappings `_meta`
information now contain a new property, called `indexTypesMap`, which
holds a list of the SO types that are part of each index.

If the list is not present, the migration logic assumes that the .kibana
index must be split (and its documents redistributed), and it forces a
reindex.

The goal of this PR is to enrich the `mappings.json` with that extra
property, so that the migrator logic will not trigger a reindex.

- Thus, these mappings must be updated whenever existing types are moved
to a separate index, or new types are added.
- Ideally, we should also update the corresponding `data.json` files, to
place each SO document to the appropriate index.

gsoldevila committed:
[dot-kibana-split] Generalise FTR and integration tests to support multiple SO indices (#154884)

Part of https://github.com/elastic/kibana/pull/154888

There are a bunch of FTR and integration tests that interact with the SO
indices directly, assuming there are only two SO indices: `.kibana` and
`.kibana_task_manager`.

The goal of this PR is to factorise that knowledge and make these tests
use a list of SO indices instead.

gsoldevila committed:
[dot-kibana-split] Update ES archiver to support multiple SO indices (#154882)

Part of https://github.com/elastic/kibana/pull/154888

ES archiver logic assumes that there are only two SO indices: '.kibana'
and '.kibana_task_manager'.
This PR updates ES archiver so that it works with a list of SO indices
instead.

---------

Co-authored-by: pgayvallet <pierre.gayvallet@elastic.co>

gsoldevila committed:
Generalise 7_13_0_transform_failures.test.ts

pgayvallet committed:
Merge remote-tracking branch 'upstream/main' into dot-kibana-split

pgayvallet committed:
updating snapshot

pgayvallet committed:
[dot-kibana-split] Adapt usages of `core.savedObjects.getKibanaIndex` to use the correct index (#155155)

### **this PR is targeting the `dot-kibana-split` feature branch and not
`main`**

## Summary

Part of https://github.com/elastic/kibana/pull/154888

In https://github.com/elastic/kibana/pull/154888, we're going to split
the `.kibana` savedObject index into multiple ones. For this reason,
calls to `core.savedObjects.getKibanaIndex` will not necessarily return
the correct value (e.g types that were moved out of this index)

This PR introduces the following SO APIs:
- `getDefaultIndex`
- `getIndexForType`
- `getIndicesForTypes`
- `getAllIndices`

And adapt plugins code to replace usages of
`core.savedObjects.getKibanaIndex` with the proper alternative

gsoldevila committed:
[dot-kibana-split] Split rules, alerts and cases saved objects into `.kibana_alerting_cases` (#154897)

## Summary

**This is targeting a feature branch**
Part of https://github.com/elastic/kibana/pull/154888

This PR aims at moving the following saved object types to a new
`.kibana_alerting_cases ` system index:
- action
- action_task_params
- alert
- api_key_pending_invalidation
- cases
- cases-comments
- cases-configure
- cases-connector-mappings
- cases-telemetry
- cases-user-actions
- connector_token
- rules-settings

---------

Co-authored-by: Christos Nasikas <christos.nasikas@elastic.co>
Co-authored-by: pgayvallet <pierre.gayvallet@elastic.co>

pgayvallet committed:
[dot-kibana-split] Split ingest types into `.kibana_ingest` (#155130)

### **This PR is targeting the `dot-kibana-split` feature branch and not
`main`**

## Summary


Part of https://github.com/elastic/kibana/pull/154888
Move the fleet/ingest saved object types to the `.kibana_ingest` index

Types moved to `.kibana_ingest`:
- epm-packages
- epm-packages-assets
- fleet-fleet-server-host
- fleet-message-signing-keys
- fleet-preconfiguration-deletion-record
- fleet-proxy
- ingest_manager_settings
- ingest-agent-policies
- ingest-download-sources
- ingest-outputs
- ingest-package-policies

---------

Co-authored-by: Gerard Soldevila <gerard.soldevila@elastic.co>

gsoldevila committed:
[dot-kibana-split] Split security solution SOs into `.kibana_security_solution` (#155129)

## Summary

**This is targeting a feature branch**
Part of https://github.com/elastic/kibana/pull/154888

This PR aims at moving the following saved object types to a new
`.kibana_security_solution` system index.

- csp-rule-template
- endpoint:user-artifact
- endpoint:user-artifact-manifest
- exception-list
- exception-list-agnostic
- osquery-manager-usage-metric
- osquery-pack
- osquery-pack-asset
- osquery-saved-query
- security-rule
- security-solution-signals-migration
- siem-detection-engine-rule-actions
- siem-ui-timeline
- siem-ui-timeline-note
- siem-ui-timeline-pinned-event

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: pgayvallet <pierre.gayvallet@elastic.co>

gsoldevila committed:
[dot-kibana-split] Split analyst experience SOs into `.kibana_analytics` (#155140)

## Summary

**This is targeting a feature branch**
Part of https://github.com/elastic/kibana/pull/154888

This PR aims at moving the following saved object types to a new
`.kibana_analytics` system index.

- canvas-element
- canvas-workpad-template
- canvas-workpad
- dashboard
- graph-workspace
- index-pattern
- kql-telemetry
- lens
- lens-ui-telemetry
- map
- search
- search-session
- search-telemetry
- url
- visualization

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: pgayvallet <pierre.gayvallet@elastic.co>

pgayvallet committed:
updating mapping files

pgayvallet committed:
fixing test message again again again

pgayvallet committed:
fix more security solution FTR helpers & data

pgayvallet committed:
fix one more test

pgayvallet committed:
Merge remote-tracking branch 'upstream/main' into dot-kibana-split

elasticmachine commented:
Pinging @elastic/kibana-core (Team:Core)

vadimkibana commented on src/plugins/share/server/url_service/saved_objects/register_url_service_saved_object_type.ts:
```diff
@@ -20,6 +21,7 @@ export const registerUrlServiceSavedObjectType = (
 ) => {
   const urlSavedObjectType: SavedObjectsType = {
     name: 'url',
+    indexPattern: ANALYTICS_SAVED_OBJECT_INDEX,
```
Does it have to by analytics index? Short URLs are use by analytics apps, but they don't have to be use only by Analytics apps, they can be used by other teams, too.

pgayvallet replied:
It doesn't have to, no.

Now, there was a mail from Rudolf to all contributors with a TTL of April, 19 (which will be yesterday) for feedback regarding the reallocation / where types should be moved.

Now that this TTL is passed, I'm not sure we still have the time to revisit it without putting the delivery of the feature even more at risk to be honest.

cc @rudolf @gsoldevila

pgayvallet replied:
Discussed about that on slack. We will note such reallocation tweaks and perform them as fixes after the merge

I'll add a section to the issue's description for that

gsoldevila replied:
I think it's a valid point, but as Pierre mentioned we can't spend a lot of time moving types around given our deadline.
I will create a PR with the requested change, to trigger CI and check for potential impact.
If there's no impact, we can attempt to merge it with the feature branch before FF.
Otherwise, updating this as a fix (post FF) seems like a viable path too.

pgayvallet committed:
rename functions

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

gsoldevila committed:
[dot-kibana-split] Address some PR remarks (#155116)

Part of https://github.com/elastic/kibana/pull/154882

This PR addresses the following NITS and remarks:
- Rename ES archiver internal methods that deal with SO indices. Some ES
archiver internal methods' names must be renamed to match their new
behaviour.
- Rename `SavedObjectsIndexPatterns` to `ALL_SAVED_OBJECT_INDICES`.

pgayvallet committed:
Merge remote-tracking branch 'upstream/main' into dot-kibana-split

mohamedhamed-ahmed reviewed (approved):
Infra changes LGTM

thomheymann reviewed (approved):
Security changes LGTM!

kpollich reviewed (approved):
Fleet changes 🚀

jbudz reviewed (approved):
`kbn-es-archiver` and `kbn-ftr-common-functional` LGTM

alisonelizabeth reviewed (approved):
Deployment Management changes LGTM (rollup plugin)

jpdjere commented on x-pack/test/detection_engine_api_integration/utils/get_rule_so_by_id.ts:
```diff
@@ -22,6 +23,6 @@ interface RuleSO {
 export const getRuleSOById = async (es: Client, id: string): Promise<SearchResponse<RuleSO>> =>
   es.search({
-    index: '.kibana',
+    index: ALERTING_CASES_SAVED_OBJECT_INDEX,
```
Nit: the JSDoc for this function still mentions the `.kibana` index. Could lead to confusion in the future

jpdjere replied:
The same is true for most modified utils in this folder

pgayvallet replied:
Fixed in https://github.com/elastic/kibana/pull/154888/commits/1409810caa538b35b988931f13caa03f5100d2cc

davismcphee reviewed (approved):
Data Discovery changes LGTM

jpdjere commented on test/functional/apps/management/_handle_version_conflict.ts:
```diff
@@ -28,7 +29,7 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
   const log = getService('log');
 
-  describe('index version conflict', function describeIndexTests() {
+  describe('FOO index version conflict', function describeIndexTests() {
```
Nit: Is this intended?

jpdjere reviewed (approved):
Security Solution Detections Rules changes - LGTM 👍 

I do however get timeouts for the migrations when starting up locally:
```
[2023-04-21T00:16:51.073+02:00][ERROR][savedobjects-service] [.kibana_analytics] Action failed with 'Request timed out'. Retrying attempt 1 in 2 seconds.
[2023-04-21T00:16:51.074+02:00][INFO ][savedobjects-service] [.kibana_analytics] CREATE_REINDEX_TEMP -> CREATE_REINDEX_TEMP. took: 120003ms.
[2023-04-21T00:16:51.075+02:00][ERROR][savedobjects-service] [.kibana_alerting_cases] Action failed with 'Request timed out'. Retrying attempt 1 in 2 seconds.
[2023-04-21T00:16:51.075+02:00][INFO ][savedobjects-service] [.kibana_alerting_cases] CREATE_REINDEX_TEMP -> CREATE_REINDEX_TEMP. took: 120003ms.
[2023-04-21T00:16:51.076+02:00][ERROR][savedobjects-service] [.kibana_ingest] Action failed with 'Request timed out'. Retrying attempt 1 in 2 seconds.
[2023-04-21T00:16:51.076+02:00][INFO ][savedobjects-service] [.kibana_ingest] CREATE_REINDEX_TEMP -> CREATE_REINDEX_TEMP. took: 120003ms.
[2023-04-21T00:16:51.076+02:00][ERROR][savedobjects-service] [.kibana_task_manager] Action failed with 'Request timed out'. Retrying attempt 1 in 2 seconds.
[2023-04-21T00:16:51.077+02:00][INFO ][savedobjects-service] [.kibana_task_manager] UPDATE_SOURCE_MAPPINGS_PROPERTIES -> UPDATE_SOURCE_MAPPINGS_PROPERTIES. took: 120003ms.
[2023-04-21T00:16:51.078+02:00][ERROR][savedobjects-service] [.kibana_security_solution] Action failed with 'Request timed out'. Retrying attempt 1 in 2 seconds.
[2023-04-21T00:16:51.078+02:00][INFO ][savedobjects-service] [.kibana_security_solution] CREATE_REINDEX_TEMP -> CREATE_REINDEX_TEMP. took: 120004ms.
[2023-04-21T00:16:51.085+02:00][ERROR][savedobjects-service] [.kibana] Action failed with 'Request timed out'. Retrying attempt 1 in 2 seconds.
```
Is this expected, or any ideas why?

pgayvallet committed:
update utils description

banderror committed:
[Security Solution] Fix flaky install_prebuilt_rules_from_real_package test (#155388)

**NOTE:** This PR will be merged to the following base PR:
https://github.com/elastic/kibana/pull/154888

## Summary

The Kibana Core team is working on splitting saved objects into multiple
indices (see the above PR). @pgayvallet said that the
`install_prebuilt_rules_from_real_package` API integration test was
incredibly flaky for them -- preventing the CI from running green:

> We had to rerun it like 4 or 5 times before being able to have a green
run on detection_engine_api_integration/security_and_space/group1
because of that specific test.

Example builds with failures:

https://buildkite.com/elastic/kibana-pull-request/builds/121292
https://buildkite.com/elastic/kibana-pull-request/builds/121246
https://buildkite.com/elastic/kibana-pull-request/builds/121092

In this PR I'm trying to fix the flake which is apparently caused by a
race condition between write and subsequent read operations that this
test performs. Please read the comments in the code for detailed
explanation.

### Checklist

- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios

pgayvallet committed:
Merge remote-tracking branch 'upstream/main' into dot-kibana-split

pgayvallet committed:
update snapshot due to merge

stratoula reviewed (approved):
Visualizations team changes LGTM

pgayvallet committed:
fixing added test

gsoldevila commented:
> I do however get timeouts for the migrations when starting up locally:
> ...
> Is this expected, or any ideas why?

@jpdjere no, this is definitely not expected. Can you share the full logs or give a bit of context of your "before" state in your local local? w

vitaliidm reviewed (approved):
Detection alerts area LGTM
Tested legacy actions migration as well, works as expected

cnasikas reviewed (approved):
ReponseOps changes LTGM!

ThomThomson reviewed (approved):
Presentation team changes LGTM!

nreese reviewed (approved):
kibana-gis changes LGTM
code review only

kevinlog reviewed (approved):
security-defend-workflows changes LGTM, we have [one PR](https://github.com/elastic/kibana/pull/155204) to remove `index: false` which would conflict with this, but we can resolve conflicts there after this PR goes in since it would be minimal.

TinaHeiligers reviewed (approved):
Awesome work!

michaelolo24 commented on x-pack/test/detection_engine_api_integration/utils/downgrade_immutable_rule.ts:
```diff
@@ -18,7 +19,7 @@ export const downgradeImmutableRule = async (
       return es.updateByQuery(
         {
-          index: '.kibana',
+          index: ALERTING_CASES_SAVED_OBJECT_INDEX,
```
why is this cases and not security?

michaelolo24 reviewed (approved):
Thanks for doing this work @gsoldevila, looks good for the timeline SO's! Just a question on whether or not QA teams have been looped in for additional testing? I'm not sure how prepared our teams are to test for any potential regressions while also managing any existing 8.8 feature work during the upcoming FF?

pgayvallet committed:
Merge remote-tracking branch 'upstream/main' into dot-kibana-split

pgayvallet commented:
> Just a question on whether or not QA teams have been looped in for additional testing?

They have been, yes.

If you think that a particular area is more at risk because of this PR, please feel free to inform us and/or QA.

darnautov reviewed (approved):
ML changes tested and LGTM

gsoldevila committed:
[dot-kibana-split] Misc SO type relocations (#155366)

## Summary

**This is targeting a feature branch**
Part of https://github.com/elastic/kibana/pull/154888

Following some remarks and feedback from the [feature branch
PR](https://github.com/elastic/kibana/pull/154888), this PR aims at
relocating some of the SO types:
- Move `'url'` type back to `.kibana`. (related
[comment](https://github.com/elastic/kibana/pull/154888#discussion_r1172367277))

pjhampton reviewed (approved):
🌔 🚀 ✨ LGTM ✨ 🚀 🌔

gsoldevila commented:
> security-defend-workflows changes LGTM, we have [one PR](https://github.com/elastic/kibana/pull/155204) to remove `index: false` which would conflict with this, but we can resolve conflicts there after this PR goes in since it would be minimal.

Thanks for letting us know! Whoever merges last should update the "dynamic" mappings.json (there's 8 of them), to get rid of the removed type `endpoint:user-artifact`.

yctercero reviewed (approved):
security-solution-platform changes LGTM. 

~~I did wonder if there are any traces of the `.siem-notification` SO. That one is part of the legacy security solution actions and didn't see it listed.~~
EDIT: It's an `alert` SO

gsoldevila committed:
Merge branch 'main' into dot-kibana-split
